src/System.Net.Sockets/src/System/Net/Sockets/_MultipleConnectAsync.cs

```diff
@@ -31,6 +39,10 @@ private enum State
 
         private object _lockObject = new object();
 
+        protected abstract bool RequiresUserConnectAttempt { get; }
```

Comment:
I've been trying to follow along with the review; I see this discussion of the PAL going on in a few different places, so I'll respond here, but it applies to the others as well.

In general, across corefx, I think we've done a good job of keeping the separation between common code and platform-specific code clean.  We've done so using a variety of patterns, as there isn't one that fits all cases.

In general, if we have a static class that needs some platform-specific functionality, we've either created a separate XxPal class that the common class calls into, or we've created partial classes to contain the static methods.  We initially started down the path of the former, but it's difficult to do well when you get to instance members and state, which pushes you down the path of the partial class design.  So the majority of corefx is on the partial class plan.

However, there have been some cases where neither of those approaches is perfectly feasible, and the best solution is to have some PAL-based capability exposed that the common code can then make decisions based on.  For example, with different file systems handling casing differently, having to duplicate functionality in the different PAL implementations based on whether the file system is case sensitive or not leads to tons of duplication.  Instead, it's much cleaner to expose a constant / readonly static from the PAL layer and let the common code branch based on that.  In most cases, the unfollowed branch doesn't even end up in the IL, as Roslyn will do dead branch elimination based on constants.  And for readonly statics like Booleans, the JIT will do dead branch elimination in the generated assembly. Another example where we hit this was in memory mapped files.  When using CreateNew, we need to create an anonymous backing store for the created map.  We'd really like to do so using an in-memory object (shm_open), but even though that function exists on both Linux and OS X, it doesn't have the charactersitics we need on OS X.  As such, we have a backup solution that is to create a file-based object.  For a variety of reasons, we're now structured so that at run time the PAL tells the common code which approach to use, and the common code branches based on that.

This is all to say that there are definitely cases where it's much cleaner IMHO to have the common code do a bit of branching based on checking capabilities exposed from the PAL.  In fact, I wish we took that path in a few more cases and avoided some code duplication that's led to some bugs.  For example, when I originally did the Unix implementation of FileStream, there were these kinds of branches, and so I ended up essentially just copying the Win32 implementation to a new file, deleting stuff that wasn't relevant, and adding in the new Unix-specific stuff.  In retrospect that was a mistake, and at some point I'd like to go revisit it.  Overtime, we've found a bug here and a bug there where because of the code copying, the implementations had drifted apart and were behaving more differently than they should have been.  When we've had to make a bug fix, we've had to do it in two different places.  Etc.

I agree it would be nice if these capability checks weren't necessary.  From a brief look over the code, though, it seems like it could be the least bad option.

My $.02.